pkg/util/audit/audit.go

```diff
@@ -0,0 +1,158 @@
+package audit
+
+import (
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/ARO-RP/pkg/util/audit/schema"
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	auditMetadataCategory    = "auditCategory"
+	auditMetadataCreatedTime = "auditCreatedTime"
+	auditMetadataFullPayload = "auditFullPayload"
+	auditMetadataLogKind     = "logKind"
+	auditMetadataOperation   = "auditOperation"
+	auditMetadataResult      = "auditResult"
+	auditMetadataSource      = "auditSource"
+	auditSourceRP            = "aro-rp"
+
+	ifxAuditCloudVer = 1.0
+	ifxAuditLogKind  = "ifxaudit"
+	ifxAuditName     = "#Ifx.AuditSchema"
+	ifxAuditVersion  = 2.1
+
+	// ifxAuditFlags is a collection of values bit-packed into a 64-bit integer.
+	// These properties describe how the event should be processed by the pipelinei
+	// in an implementation-independent way.
+	ifxAuditFlags = 257
+
+	// ifxAuditPopSample is the effective percentage sample rate for this event at the
+	// time it was generated by a client (either the default value specified in
+	// the event definition, system default or a dynamic override from a cloud
+	// settings service).
+	// not used by Azure
+	ifxAuditPopSample = 0.0
+)
+
+var (
+	// epoch is an unique identifier associated with the current session of the
+	// telemetry library running on the platform. It must be stable during a
+	// session, and has no implied ordering across sessions.
+	epoch string
+
+	// sequence field used to track absolute order of uploaded events, per session.
+	// it is reset when the ARO component is restarted.
+	seqNum      uint64
+	seqNumMutex *sync.Mutex
+)
+
+func init() {
+	epoch = uuid.NewV4().String()
+	seqNum = 1
+	seqNumMutex = new(sync.Mutex)
+}
+
+// EmitRPLog is used by aro-rp to emit audit logs.
+// Caller is responsible for providing the logrus.Entry object that contains
+// the appropriate formatter, hooks, log level etc.
+// All audit logs will be emitted at the INFO level.
+func EmitRPLog(entry logrus.Entry, log *Log) error {
+	return emitLog(entry, log, auditSourceRP)
+}
+
+func emitLog(entry logrus.Entry, log *Log, source string) error {
+	payload := log.toPayload()
+
+	fields := logrus.Fields{}
+	fields[auditMetadataCategory] = payload.Category
+	fields[auditMetadataCreatedTime] = time.Now().UTC().Format(time.RFC3339)
+	fields[auditMetadataLogKind] = ifxAuditLogKind
+	fields[auditMetadataOperation] = payload.OperationName
+	fields[auditMetadataResult] = payload.Result.ResultType
+	fields[auditMetadataSource] = source
+
+	marshalled, err := marshal(payload)
+	if err != nil {
+		return err
+	}
+	fields[auditMetadataFullPayload] = marshalled
```

Comment:
This field captures the entire audit log "envelope" as JSON, as defined by the schema.